sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingInterpreter.scala

```diff
@@ -0,0 +1,82 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.scripting
+
+import org.apache.spark.sql.catalyst.analysis.UnresolvedIdentifier
+import org.apache.spark.sql.catalyst.parser.{CompoundBody, CompoundPlanStatement, SingleStatement}
+import org.apache.spark.sql.catalyst.plans.logical.{CreateVariable, DropVariable, LogicalPlan}
+import org.apache.spark.sql.catalyst.trees.Origin
+
+/**
+ * Trait for SQL Scripting interpreters.
+ */
+trait ProceduralLanguageInterpreter {
+  /**
+   * Build execution plan and return statements that need to be executed,
+   *  wrapped in the execution node.
+   * @param compound CompoundBody for which to build the plan.
+   * @return Iterator through collection of statements to be executed.
+   */
+  def buildExecutionPlan(compound: CompoundBody) : Iterator[CompoundStatementExec]
+}
+
+/**
+ * Concrete implementation of the interpreter for SQL scripting.
+ */
+case class SqlScriptingInterpreter() extends ProceduralLanguageInterpreter {
+  /** @inheritdoc */
+  override def buildExecutionPlan(compound: CompoundBody): Iterator[CompoundStatementExec] = {
+    transformTreeIntoExecutable(compound).asInstanceOf[CompoundBodyExec]
+  }
+
+  /**
+   * Fetch the name of the Create Variable plan.
+   * @param plan Plan to fetch the name from.
+   * @return Name of the variable.
+   */
+  private def getDeclareVarNameFromPlan(plan: LogicalPlan): Option[UnresolvedIdentifier] =
+    plan match {
+      case CreateVariable(name: UnresolvedIdentifier, _, _) => Some(name)
+      case _ => None
+    }
+
+  /**
+   * Transform the parsed tree to the executable node.
+   * @param node Root node of the parsed tree.
+   * @return Executable statement.
+   */
+  private def transformTreeIntoExecutable(node: CompoundPlanStatement): CompoundStatementExec =
+    node match {
+      case body: CompoundBody =>
+        val variables = body.collection.flatMap {
+          case st: SingleStatement => getDeclareVarNameFromPlan(st.parsedPlan)
+          case _ => None
+        }
+        val dropVariables = variables
+          .map(varName => DropVariable(varName, ifExists = true))
+          .map(new SingleStatementExec(_, Origin(), isInternal = true))
+          .reverse
+        new CompoundBodyExec(
+          body.collection.map(st => transformTreeIntoExecutable(st)) ++ dropVariables)
```

Comment:
Instead of just concatenating DROPs alternative could be to:
1) Keep "cleanup/internal" statements in side structure.
2) Execute them as part of `finalize` or `reset` calls.

My fear is that by just keeping them at the end it might be hard to support following scenarios:
1) premature return (e.g. during break)
2) error handling (exception thrown in the middle of the block).